New script actions for enabling & disabling AI triggers

FS-21 opened the pull request:
Ai trigger manipulation. More information in the documentation

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **New Features**
	- Introduced new script actions for AI trigger management, including setting indices and enabling/disabling triggers based on various criteria.
	- Enhanced functionality for managing AI triggers in game scenarios.
	- Added automatic deletion logic for passengers in vehicles.

- **Bug Fixes**
	- Resolved issues in the MC deployer functionality.

- **Documentation**
	- Updated AI scripting and mapping guidance.
	- Documented new and enhanced logics affecting game mechanics.
	- Added contributions to the help documentation.

- **Refactor**
	- Improved data handling and processing of AI triggers in the game's rules system.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

FS-21 committed:
New actions 127 to 133

Ai trigger manipulation. More information in the documentation

FS-21 committed:
I forgot including in this branch some code

FS-21 committed:
Merge branch 'develop' into feature/ai-trigger-manipulation-actions

FS-21 commented:
These triggers were tested by me with and currently I haven't found any new bug.

FS-21 committed:
Merge branch 'develop' into feature/ai-trigger-manipulation-actions

FS-21 committed:
Merge branch 'develop' into feature/ai-trigger-manipulation-actions

FS-21 committed:
Point to latest YRpp

Metadorius committed:
Update GH Actions workflows
- Split non-PR and main branch nightly build PRs
- Use PR HEAD instead of merge commit

FS-21 committed:
Merge remote-tracking branch 'origin/develop' into feature/ai-trigger-manipulation-actions

FS-21 committed:
Fix name due to YRpp change

FS-21 committed:
Merge remote-tracking branch 'origin/develop' into feature/ai-trigger-manipulation-actions

FS-21 committed:
Merge remote-tracking branch 'origin/develop' into feature/ai-trigger-manipulation-actions

FS-21 committed:
Merge remote-tracking branch 'origin/develop' into feature/ai-trigger-manipulation-actions

FS-21 committed:
Merge remote-tracking branch 'origin/develop' into feature/ai-trigger-manipulation-actions

FS-21 committed:
Merge branch 'develop' into feature/ai-trigger-manipulation-actions

FS-21 committed:
Point to latest YRpp

FS-21 committed:
Merge remote-tracking branch 'origin/develop' into feature/ai-trigger-manipulation-actions

FS-21 committed:
Point changelog to v0.3.1 and not v0.3

FS-21 committed:
Actions renumeration

Changed action numbers:

SetSideIdxForManagingTriggers = 127 -> 16005,
SetHouseIdxForManagingTriggers = 128 -> 16006,
ManageAllAITriggers = 129 -> 16007,
EnableTriggersFromList = 130 -> 16008,
DisableTriggersFromList = 131 -> 16009,
DisableTriggersWithObjects = 133 -> 16010,
EnableTriggersWithObjects = 132 -> 16011,

FS-21 committed:
Added lost documentation from older merges

And updated the action ranges of these new actions

FS-21 committed:
small change in docs

FS-21 committed:
Merge branch 'develop' into feature/ai-trigger-manipulation-actions

FS-21 committed:
Compilation fixes

FS-21 committed:
Merge branch 'develop' into feature/ai-trigger-manipulation-actions

FS-21 committed:
.

FS-21 committed:
Merge branch 'develop' into feature/ai-trigger-manipulation-actions

FS-21 committed:
Merge branch 'develop' into feature/ai-trigger-manipulation-actions

FS-21 committed:
Merge branch 'develop' into feature/ai-trigger-manipulation-actions

FS-21 committed:
Merge branch 'develop' into feature/ai-trigger-manipulation-actions

FS-21 committed:
Moved code into the new file

and more tweaks

FS-21 committed:
Merge branch 'develop' into feature/ai-trigger-manipulation-actions

Starkku commented on docs/New-or-Enhanced-Logics.md:
```diff
@@ -1305,7 +1305,10 @@ FeedbackWeapon=  ; WeaponType
 - In addition to allowing custom radiation types, several enhancements are also available to the default radiation type defined in `[Radiation]`, such as ability to set owner & invoker or deal damage against buildings. See [Custom Radiation Types](#custom-radiation-types) for more details.
 
-### Strafing aircraft weapon customization
+### `500 - 523` Edit Variable
+- Operate a variable's value
+    - The variable's value type is int16 instead of int32 in trigger actions for some reason, which means it ranges from -2^15 to 2^15-1.
+        - Any numbers exceeding this limit will lead to unexpected results!
```
This stuff doesn't belong to this page, probably something to fix by rebasing the branch or whatever.

FS-21 replied:
I think I fixed this, seems an old leftover or a bad merge because I never rebase.

Starkku commented on src/Ext/House/Body.cpp:
```diff
@@ -480,6 +480,154 @@ int HouseExt::ExtData::CountOwnedPresentAndLimboed(TechnoTypeClass* pTechnoType)
+	// Transtale the Multiplayer index into a valid index for the HouseClass array
+	if (param >= HouseClass::PlayerAtA && param <= HouseClass::PlayerAtH)
+	{
+		switch (param)
```
Does not need an entire switch case, deduct value of `HouseClass::PlayerAtA` from the param and you get what you need.

FS-21 replied:
I'll upload a remake of that function applying the feedback

Starkku commented on src/Ext/House/Body.cpp:
```diff
@@ -480,6 +480,154 @@ int HouseExt::ExtData::CountOwnedPresentAndLimboed(TechnoTypeClass* pTechnoType)
+	}
+
+	// Special cases
+	switch (param)
```
There's a fair bit of code repetition across this switch case, if you can find a way to condense it without slowing the code down too much with heavily stacked if statements that'd be great.

Starkku commented on src/Ext/Script/Body.ManageAITriggers.cpp:
```diff
@@ -0,0 +1,195 @@
+
+void ScriptExt::ManageTriggersFromList(TeamClass* pTeam, int idxAITriggerType = -1, bool isEnabled = false)
+{
+	if (!pTeam)
+		return;
```
This kind of `nullptr` check is really not needed. Generally speaking if the pointer of 'this' or 'current' object is null, you have a much bigger fish to fry than your code crashing (in fact, it is probably better to let it crash than have it obscure a potential issue).

Starkku commented on src/Ext/Script/Body.ManageAITriggers.cpp:
```diff
@@ -0,0 +1,195 @@
+	if (idxAITriggerType < 0)
+		idxAITriggerType = pScript->Type->ScriptActions[pScript->CurrentMission].Argument;
+
+	if (idxAITriggerType < 0)
+		return;
+
+	if (RulesExt::Global()->AITriggersLists.size() <= 0)
+		return;
```
Condense to single if statement.

Starkku commented on src/Ext/Script/Body.ManageAITriggers.cpp:
```diff
@@ -0,0 +1,195 @@
+
+void ScriptExt::ManageAllTriggersFromHouse(TeamClass* pTeam, HouseClass* pHouse = nullptr, int sideIdx = -1, int houseIdx = -1, bool isEnabled = true)
+{
+	if (!pTeam)
+		return;
```
Ditto with the `nullptr` check here.

Starkku commented on src/Ext/Script/Body.ManageAITriggers.cpp:
```diff
@@ -0,0 +1,195 @@
+
+void ScriptExt::SetSideIdxForManagingTriggers(TeamClass* pTeam, int sideIdx = -1)
+{
+	if (!pTeam)
+		return;
```
Ditto with the `nullptr` check here.

Starkku commented on src/Ext/Script/Body.ManageAITriggers.cpp:
```diff
@@ -0,0 +1,195 @@
+	if (sideIdx < 0)
+		sideIdx = pScript->Type->ScriptActions[pScript->CurrentMission].Argument;
+
+	if (sideIdx < -1)
+		sideIdx = -1;
```
Is there a reason it would ever be below -1 and does it explicitly check for -1 somewhere as opposed to < 0?

FS-21 replied:
As I menmtioned before for houses, in triggers houses and sides -1 value means "any/not specified" so if you made the mistake it fix the value to -1 and continue the execution.

Starkku commented on src/Ext/Script/Body.ManageAITriggers.cpp:
```diff
@@ -0,0 +1,195 @@
+
+void ScriptExt::SetHouseIdxForManagingTriggers(TeamClass* pTeam, int houseIdx = 1000000)
+{
+	if (!pTeam)
+		return;
```
Ditto with the `nullptr` check here.

Starkku commented on src/Ext/Script/Body.ManageAITriggers.cpp:
```diff
@@ -0,0 +1,195 @@
+
+	auto pScript = pTeam->CurrentScript;
+
+	if (houseIdx == 1000000)
+		houseIdx = pScript->Type->ScriptActions[pScript->CurrentMission].Argument;
```
What's this magic number?

FS-21 replied:
A default value very far from the possible game values.

Starkku commented on src/Ext/Script/Body.ManageAITriggers.cpp:
```diff
@@ -0,0 +1,195 @@
+
+	houseIdx = HouseExt::GetHouseIndex(houseIdx, pTeam, nullptr);
+
+	if (houseIdx < -1)
+		houseIdx = -1;
```
Is there a reason it would ever be below -1 and does it explicitly check for -1 somewhere as opposed to < 0?

FS-21 replied:
Sanity check:
A crazy modder/mapper could put a invalid negative value so by default is set to "not specified" (-1 index).

Starkku commented on src/Ext/Script/Body.ManageAITriggers.cpp:
```diff
@@ -0,0 +1,195 @@
+
+void ScriptExt::ManageAITriggers(TeamClass* pTeam, int enabled = -1)
+{
+	if (!pTeam)
+		return;
```
Ditto with the `nullptr` check here.

Starkku commented on src/Ext/Script/Body.ManageAITriggers.cpp:
```diff
@@ -0,0 +1,195 @@
+		return;
+
+	auto pTeamData = TeamExt::ExtMap.Find(pTeam);
+	if (!pTeamData)
+	{
+		// This action finished
+		pTeam->StepCompleted = true;
+	}
```
Put the comment after the line and remove the brackets around the if statement.

FS-21 replied:
This is a bug, if this ended then it should "return" :-S
Mainly because after that if statement we still work with "pTeamData->..."

Starkku commented on src/Ext/Script/Body.ManageAITriggers.cpp:
```diff
@@ -0,0 +1,195 @@
+
+void ScriptExt::ManageTriggersWithObjects(TeamClass* pTeam, int idxAITargetType = -1, bool isEnabled = false)
+{
+	if (!pTeam)
+		return;
```
Ditto with the `nullptr` check here.

Starkku commented on src/Ext/Script/Body.ManageAITriggers.cpp:
```diff
@@ -0,0 +1,195 @@
+	if (idxAITargetType < 0)
+		idxAITargetType = pScript->Type->ScriptActions[pScript->CurrentMission].Argument;
+
+	if (idxAITargetType < 0)
+		return;
+
+	if (RulesExt::Global()->AITargetTypesLists.size() <= 0)
+		return;
```
Condense to single if statement.

Starkku commented on src/Ext/Script/Body.ManageAITriggers.cpp:
```diff
@@ -0,0 +1,195 @@
+	{
+		DynamicVectorClass<TechnoTypeClass*> entriesList;
+
+		if (pTrigger->Team1)
+		{
+			for (auto entry : pTrigger->Team1->TaskForce->Entries)
+			{
+				if (entry.Amount > 0)
+					entriesList.AddItem(entry.Type);
+			}
+		}
+
+		if (pTrigger->Team2)
+		{
+			for (auto entry : pTrigger->Team2->TaskForce->Entries)
+			{
+				if (entry.Amount > 0)
+					entriesList.AddItem(entry.Type);
+			}
+		}
```
Could consider using lambda expression here and only write the code once given that the only difference is the TeamType used.

FS-21 replied:
I wrote this lambda:

```
auto AddTechnosFromTeam = [&](DynamicVectorClass<TechnoTypeClass*>& entriesList, TeamTypeClass* pTeam)
{
	if (pTeam)
	{
		for (auto entry : pTeam->TaskForce->Entries)
		{
			if (entry.Amount > 0)
				entriesList.AddItem(entry.Type);
		}
	}
};
```

I'm not familiar with lambdas so if I made a mistake or it needs improvement feel free to tell me.

Starkku reviewed (changes requested):
Mostly code style / verbosity stuff like unnecessary checks and duplication of code that should probably be addressed.

Starkku commented on src/Ext/Script/Body.ManageAITriggers.cpp:
```diff
@@ -0,0 +1,195 @@
+			}
+		}
+
+		if (entriesList.Count > 0)
+		{
+			for (auto entry : entriesList)
+			{
+				if (objectsList.FindItemIndex(entry) >= 0)
+				{
+					pTrigger->IsEnabled = isEnabled;
+					break;
+				}
+			}
+		}
```
Also don't need to check for count if you're using for-each loop.

FS-21 replied:
Good point!

FS-21 committed:
Rewrote HouseExt::GetHouseIndex(...) with the requested changes.
Applied all the feedback to these AITrigger manipulation functions. I also tweaked some "return" that didn't explain "why" and avoided an infinite loop.

Tweaked ScriptExt::SetCloseEnoughDistance() with a small change not related to this PR (is too small for a PR and isn't a bug).

FS-21 commented:
I rewrote HouseExt::GetHouseIndex(...) with the requested changes.
Applied all the feedback to these AITrigger manipulation functions. I also tweaked some "return" that didn't explain "why" and avoided an infinite loop.

I also tweaked ScriptExt::SetCloseEnoughDistance() with a small change not related to this PR (is too small for a PR and isn't a bug).

I still have to touch that documentation mistake :-/

FS-21 commented:
I didn't notice that 16005 has been occupied by another script Action so I'll move the number to the first emply.

FS-21 committed:
Fix small mistake

FS-21 committed:
Moved scriptType Action

Before: 16005
After: 16012
Reason: Another one was merged and took the 16005

FS-21 committed:
Merge branch 'develop' into feature/ai-trigger-manipulation-actions

FS-21 committed:
Small changes

Due to review feedback

FS-21 committed:
Merge remote-tracking branch 'origin/develop' into feature/ai-trigger-manipulation-actions

FS-21 committed:
FA2 documentation

FS-21 committed:
fix mistake

FS-21 committed:
Merge branch 'develop' into feature/ai-trigger-manipulation-actions

FS-21 committed:
Merge branch 'develop' into feature/ai-trigger-manipulation-actions

FS-21 committed:
Merge branch 'develop' into feature/ai-trigger-manipulation-actions

FS-21 committed:
Merge branch 'develop' into feature/ai-trigger-manipulation-actions

FS-21 committed:
Merge branch 'develop' into feature/ai-trigger-manipulation-actions

FS-21 committed:
Merge branch 'develop' into feature/ai-trigger-manipulation-actions

FS-21 committed:
Merge branch 'develop' into feature/ai-trigger-manipulation-actions

FS-21 committed:
Merge branch 'develop' into feature/ai-trigger-manipulation-actions

FS-21 committed:
Merge branch 'develop' into feature/ai-trigger-manipulation-actions

FS-21 committed:
Merge branch 'develop' into feature/ai-trigger-manipulation-actions

FS-21 committed:
Merge branch 'develop' into feature/ai-trigger-manipulation-actions
